### PR TITLE
Capture CUDA logging output

### PR DESCRIPTION
**Pull request by fraxy-v**
Addresses logging capture issues reported here #5797.

This PR adds a CUDA logging callback similar to that in `ggml-metal.h`:
https://github.com/ggerganov/llama.cpp/blob/583fd6b000ec9ad1b465b5c98524f4a0ae388077/ggml-metal.h#L43

However, the issue OP has certainly experienced other uncaptured log messages. For example:
```
MY LOGGER SAYS: ggml_cuda_host_malloc: failed to allocate 256.00 MiB of pinned memory: no CUDA-capable device is detected
MY LOGGER SAYS: llama_kv_cache_init:        CPU KV buffer size =   256.00 MiB
MY LOGGER SAYS: llama_new_context_with_model: KV self size  =  256.00 MiB, K (f16):  128.00 MiB, V (f16):  128.00 MiB
MY LOGGER SAYS: ggml_cuda_host_malloc: failed to allocate 0.12 MiB of pinned memory: no CUDA-capable device is detected
MY LOGGER SAYS: llama_new_context_with_model:        CPU  output buffer size =     0.12 MiB
ggml_gallocr_reserve_n: reallocating CUDA_Host buffer from size 0.00 MiB to 70.50 MiB
MY LOGGER SAYS: ggml_cuda_host_malloc: failed to allocate 70.50 MiB of pinned memory: no CUDA-capable device is detected
MY LOGGER SAYS: llama_new_context_with_model:  CUDA_Host compute buffer size =    70.50 MiB
MY LOGGER SAYS: llama_new_context_with_model: graph nodes  = 1030
MY LOGGER SAYS: llama_new_context_with_model: graph splits = 1

system_info: n_threads = 4 / 8 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 | 
sampling: 
	repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
	top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
	mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
sampling order: 
CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temperature 
generate: n_ctx = 512, n_batch = 2048, n_predict = -1, n_keep = 1


<s> Tell me a story about a llama
MY LOGGER SAYS: 
MY LOGGER SAYS: llama_print_timings:   
```
As evident methods in `ggml-alloc.c`, e.g. `ggml_gallocr_reserve_n`, use directly `stderr`. So that begs the question whether the internal logging in `llama.cpp` should not be made commonly available, e.g. in `common.h`.

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
> As evident methods in ggml-alloc.c, e.g. ggml_gallocr_reserve_n, use directly stderr. So that begs the question whether the internal logging in llama.cpp should not be made commonly available, e.g. in common.h.

Most of the printfs there are during Debug builds - these are not a problem.
For the rest though, I'm not sure - I am wondering if we should just keep using `fprintf(stderr, ...` for errors and only propagate a log function when we want to print some info/warning messages (such as the ones addressed in the current issue).

File: ggml-cuda.cu [hunk 1]
```diff
@@ -43,19 +43,64 @@
...
 
+static void ggml_cuda_default_log_callback(enum ggml_log_level level, const char * msg, void * user_data) {
+    GGML_UNUSED(user_data);
+    if (level == GGML_LOG_LEVEL_WARN) {
+        fprintf(stderr, "warning: %s", msg);
+    } else if (level == GGML_LOG_LEVEL_ERROR) {
+        fprintf(stderr, "error: %s", msg);
+    } else {
+        fprintf(stderr, "%s", msg);
+    }
```
Comment (slaren) [inline ggml-cuda.cu:61]: Do not add these prefixes.

File: ggml-cuda.cu [hunk 2]
```diff
@@ -43,19 +43,64 @@
...
 void ggml_cuda_error(const char * stmt, const char * func, const char * file, int line, const char * msg) {
     int id = -1; // in case cudaGetDevice fails
     cudaGetDevice(&id);
 
-    fprintf(stderr, "CUDA error: %s\n", msg);
-    fprintf(stderr, "  current device: %d, in function %s at %s:%d\n", id, func, file, line);
-    fprintf(stderr, "  %s\n", stmt);
+    GGML_CUDA_LOG_INFO("CUDA error: %s\n", msg);
+    GGML_CUDA_LOG_INFO("  current device: %d, in function %s at %s:%d\n", id, func, file, line);
+    GGML_CUDA_LOG_INFO("  %s\n", stmt);
```
Comment (slaren) [inline ggml-cuda.cu:103]: These should be printed as errors.

File: llama.cpp [hunk 3]
```diff
@@ -18036,6 +18036,8 @@ void llama_log_set(ggml_log_callback log_callback, void * user_data) {
     g_state.log_callback_user_data = user_data;
 #ifdef GGML_USE_METAL
     ggml_backend_metal_log_set_callback(g_state.log_callback, g_state.log_callback_user_data);
+#elif GGML_USE_CUDA
+    ggml_backend_cuda_log_set_callback(g_state.log_callback, g_state.log_callback_user_data);
```
Comment (slaren) [inline llama.cpp:18040]: This also needs to be called during the `llama_state` constructor.

File: llama.cpp [hunk 4]
```diff
@@ -1778,6 +1778,8 @@ struct llama_state {
     llama_state() {
 #ifdef GGML_USE_METAL
         ggml_backend_metal_log_set_callback(log_callback, log_callback_user_data);
+#elif GGML_USE_CUDA
```
Comment (slaren) [inline llama.cpp:1781]:
```suggestion
#elif defined(GGML_USE_CUDA)
```

File: llama.cpp [hunk 5]
```diff
@@ -18036,6 +18038,8 @@ void llama_log_set(ggml_log_callback log_callback, void * user_data) {
     g_state.log_callback_user_data = user_data;
 #ifdef GGML_USE_METAL
     ggml_backend_metal_log_set_callback(g_state.log_callback, g_state.log_callback_user_data);
+#elif GGML_USE_CUDA
```
Comment (slaren) [inline llama.cpp:18041]:
```suggestion
#elif defined(GGML_USE_CUDA)
```

**Review by slaren — APPROVED** [review]

**Pull request merged**